src/components/tl/mlx5/mcast/tl_mlx5_mcast_progress.c

```diff
@@ -402,18 +403,29 @@ ucc_status_t ucc_tl_mlx5_mcast_process_packet(ucc_tl_mlx5_mcast_coll_comm_t *com
 
     if (pp->length > 0 ) {
         dest = req->ptr + PSN_TO_RECV_OFFSET(pp->psn, req, comm);
-        
         if (comm->cuda_mem_enabled) {
-            mem_type = UCC_MEMORY_TYPE_CUDA;
-        } else {
-            mem_type = UCC_MEMORY_TYPE_HOST;
-        }
+            while (req->exec_task != NULL) {
+                EXEC_TASK_TEST("failed to complete the nb memcpy", req->exec_task, comm->lib);
+            }
 
-        status = ucc_mc_memcpy(dest, (void*) pp->buf, pp->length,
-                               mem_type, mem_type);
-        if (ucc_unlikely(status != UCC_OK)) {
-            tl_error(comm->lib, "failed to copy buffer");
-            return status;
+            /* for cuda memcpy use nonblocking copy */
+            status = ucc_coll_task_get_executor(req->coll_task, &exec);
+            if (ucc_unlikely(status != UCC_OK)) {
+                return status;
+            }
+
+            eargs.task_type = UCC_EE_EXECUTOR_TASK_COPY;
+            eargs.copy.src  = (void*) pp->buf;
+            eargs.copy.dst  = dest;
+            eargs.copy.len  = pp->length;
+
+            assert(req->exec_task == NULL);
+            status = ucc_ee_executor_task_post(exec, &eargs, &req->exec_task);
+            if (ucc_unlikely(status != UCC_OK)) {
+                return status;
+            }
+        } else {
```

Comment:
condition is not needed, executor can copy host buffers as well

Reply:
That is a good question. However, it seems like i had to use ```schedule``` if i wanted to use executors for a task. Please look at `ucc_coll_task_get_executor()` function. As you can see, we need to have ```task->schedule``` ready in order to obtain a ```task->executor```.

Reply:
It will not be nonblocking in case of host memory. Do you see perf difference when you use executor vs regular memcpy?

Reply:
oh this is good to know that for host copy it is already blocking, I see not much performance diff as well, I will make this changes accordingly.